### PR TITLE
273 migrate to terriajs

### PR DESCRIPTION
**Pull request by LukeParky**
DESCRIPTION OF PR:

Closes:
* #273 
* #162 
* #254 
* #255 
* #155 
* #148 
  (courtesy of PR #284)

Migrate to using TerriaJS as the primary front-end, enabling MEDUSA front end controls.
This requires the use of OGC WebProcessingService (WPS) to run processing tasks.

A new EnvVariable, `ROOF_SURFACE_DATASET_PATH` is added to `.env.template`. A new issue #283 will add this into git so that it doesn't need to be copied from network storage.

Another issue #280 is for removing the old front-end.

This PR reduces the size of the AOI, #282 will extend it again

## Developer Checklist
- [X] Make code change
- [ ] Update tests
  - [ ] Update / create new tests
  - [ ] Ensure these tests have the expected behaviour
  - [ ] Test locally and ensure tests are passing
- [X] Update documentation
  - [X] Readme
  - [X] Docstrings
  - [X] Comments
  - [ ] Wiki

## Reviewer Checklist
- [ ] Check new code for code smells
- [ ] Check new tests
  - [ ] Ensure adequate coverage
  - [ ] Check for code smells within tests
- [ ] Check if documentation needs updating
  - [ ] Readme
  - [ ] Docstrings
  - [ ] Comments
  - [ ] Wiki

### CODE REVIEW

File: src/pywps.cfg [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+contact_url=http://pywps.org
+contact_hours=8:00-20:00UTC
+contact_instructions=Knock on the door
+contact_role=pointOfContact
+
+[server]
+maxsingleinputsize=1mb
+maxrequestsize=3mb
+url=http://localhost:5000/wps
+outputurl=http://localhost:5000/outputs/
```
Comment (LukeParky) [inline src/pywps.cfg:29]: Update this entire file

File: terriajs/client_config.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
+  "initializationUrls": [
+    "catalog"
+  ],
+  "parameters": {
+    "appName": "Ōtākaro Digital Twin",
+    "supportEmail": "luke.parkinson@canterbury.ac.nz",
+    "cesiumIonAccessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJqdGkiOiI4NDE2NGZmYy01MDRjLTQyOWUtYjdlYS05M2RmNTMyMGQzYzIiLCJpZCI6Nzk4NzUsImlhdCI6MTY0MjQ2MzEzNX0.bajvCxkIy698jp_y72DmHCqwN963H7yHmOoHLaxSTIU",
```
Comment (LukeParky) [inline terriajs/client_config.json:8]: Use environment variables here (this key has already been disabled)

Comment (LukeParky) [reply terriajs/client_config.json:8]: Or remove the need for it

File: Dockerfile
Comment (LukeParky) [inline Dockerfile:1]: Update README with additional instructions for copying gdb file while waiting on approval from data owners

**Comment by LukeParky** [comment]
@caseyli92 this is essentially ready for a code review but additional documentation is required to enable you to easily test it. (you need the .gdb in the correct folder, will add more documentation around this shortly)

File: environment.yml [hunk 3]
```diff
@@ -6,7 +6,7 @@ channels:
   - conda-forge
   - defaults
 dependencies:
-  - conda-pack>=0.7.1
+  - conda-pack==0.7.1
```
Comment (caseyli92) [inline environment.yml:9]: Could you please add a comment for this change?

File: src/app.py [hunk 4]
```diff
@@ -98,6 +108,20 @@ def index() -> Response:
     """, OK)
 
 
+@app.route('/wps', methods=['GET', 'POST'])
```
Comment (caseyli92) [inline src/app.py:111]: The SonarCloud hotspot risk needs to be resolved.

File: src/digitaltwin/data_to_db.py [hunk 5]
```diff
@@ -425,3 +433,62 @@ def user_log_info_to_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> Non
...
+        # This data has the deeplearn_matclass with roof types we need
+        log.info(f"Reading roof surface points from {EnvVariable.ROOF_SURFACE_DATASET_PATH}.")
+        roof_surface_points = gpd.read_file(EnvVariable.ROOF_SURFACE_DATASET_PATH,
+                                            layer="CCC_Lynker_RoofMaterial_Update_2023")
+        # Remove rows of building_Id and deeplearn_subclass that are NANs
+        roof_surface_points = roof_surface_points.dropna(subset=['building_Id', 'deeplearn_subclass'])
+        # Store the building_point_data to the database table
+        log.info("Adding roof_surface_points table to the database.")
+        roof_surface_points.to_postgis("roof_surface_points", engine, index=False, if_exists="replace")
+
```
Comment (caseyli92) [inline src/digitaltwin/data_to_db.py:461]:
```suggestion
        log.info("Adding 'roof_surface_points' to the database.")
        roof_surface_points.to_postgis("roof_surface_points", engine, index=False, if_exists="replace")
        log.info("Successfully added 'roof_surface_points' to the database.")
```

File: src/digitaltwin/data_to_db.py [hunk 6]
```diff
@@ -425,3 +433,62 @@ def user_log_info_to_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> Non
...
+    # Check if the table already exist in the database
+    if check_table_exists(engine, "roof_surface_polygons"):
+        log.info("roof_surface_polygons data already exists in the database.")
+    else:
+        # Read roof surface polygons from outside
+        log.info(f"Reading roof surface polygons file {EnvVariable.ROOF_SURFACE_DATASET_PATH}.")
+        roof_surface_polygons = gpd.read_file(EnvVariable.ROOF_SURFACE_DATASET_PATH, layer="BuildingPolygons")
+        # Store the building_point_data to the database table
+        log.info("Adding roof_surface_polygons table to the database.")
+        roof_surface_polygons.to_postgis("roof_surface_polygons", engine, index=False, if_exists="replace")
```
Comment (caseyli92) [inline src/digitaltwin/data_to_db.py:481]:
```suggestion
        log.info("Adding 'roof_surface_polygons' to the database.")
        roof_surface_polygons.to_postgis("roof_surface_polygons", engine, index=False, if_exists="replace")
        log.info("Successfully added 'roof_surface_polygons' to the database.")
```

File: src/digitaltwin/data_to_db.py [hunk 7]
```diff
@@ -425,3 +433,62 @@ def user_log_info_to_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> Non
     query = UserLogInfo(source_table_list=table_list, geometry=catchment_geom)
     # Execute the query
     execute_query(engine, query)
+
+
+def save_roof_surface_type_points_to_db(engine: Engine) -> None:
+    """
+    Read roof surface type data, then store them into database.
```
Comment (caseyli92) [inline src/digitaltwin/data_to_db.py:440]:
```suggestion
    """
    Read roof surface type points data then store them into database.
```

File: src/digitaltwin/data_to_db.py [hunk 8]
```diff
@@ -425,3 +433,62 @@ def user_log_info_to_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> Non
...
+        # Remove rows of building_Id and deeplearn_subclass that are NANs
+        roof_surface_points = roof_surface_points.dropna(subset=['building_Id', 'deeplearn_subclass'])
+        # Store the building_point_data to the database table
+        log.info("Adding roof_surface_points table to the database.")
+        roof_surface_points.to_postgis("roof_surface_points", engine, index=False, if_exists="replace")
+
+
+def save_roof_surface_polygons_to_db(engine: Engine) -> None:
+    """
+    Read roof surface geometry data then store them into database.
```
Comment (caseyli92) [inline src/digitaltwin/data_to_db.py:465]:
```suggestion
    """
    Read roof surface polygons data then store them into database.
```

File: src/digitaltwin/data_to_db.py [hunk 9]
```diff
@@ -425,3 +433,62 @@ def user_log_info_to_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> Non
...
+    Read roof surface type data, then store them into database.
+
+    Parameters
+    ----------
+    engine : Engine
+        The engine used to connect to the database.
+    """
+    # Check if the table already exist in the database
+    if check_table_exists(engine, "roof_surface_points"):
+        log.info("roof_surface_points data already exists in the database.")
```
Comment (caseyli92) [inline src/digitaltwin/data_to_db.py:449]:
```suggestion
    if check_table_exists(engine, "roof_surface_points"):
        log.info("'roof_surface_points' data already exists in the database.")
```

File: src/digitaltwin/data_to_db.py [hunk 10]
```diff
@@ -425,3 +433,62 @@ def user_log_info_to_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> Non
...
+    Read roof surface geometry data then store them into database.
+
+    Parameters
+    ----------
+    engine : Engine
+        The engine used to connect to the database.
+    """
+    # Check if the table already exist in the database
+    if check_table_exists(engine, "roof_surface_polygons"):
+        log.info("roof_surface_polygons data already exists in the database.")
```
Comment (caseyli92) [inline src/digitaltwin/data_to_db.py:474]:
```suggestion
    if check_table_exists(engine, "roof_surface_polygons"):
        log.info("'roof_surface_polygons' data already exists in the database.")
```

File: src/digitaltwin/data_to_db.py [hunk 11]
```diff
@@ -425,3 +433,62 @@ def user_log_info_to_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> Non
...
+        roof_surface_polygons = gpd.read_file(EnvVariable.ROOF_SURFACE_DATASET_PATH, layer="BuildingPolygons")
+        # Store the building_point_data to the database table
+        log.info("Adding roof_surface_polygons table to the database.")
+        roof_surface_polygons.to_postgis("roof_surface_polygons", engine, index=False, if_exists="replace")
+
+
+def save_roof_surface_type_data_to_db(engine: Engine) -> None:
+    """
+    Read roof surface type dataset file and save data into database.
+
```
Comment (caseyli92) [inline src/digitaltwin/data_to_db.py:487]:
```suggestion
    """
    Read roof surface type points and polygons from the data file and store them in the database.
```

File: src/digitaltwin/data_to_db.py [hunk 12]
```diff
@@ -425,3 +433,62 @@ def user_log_info_to_db(engine: Engine, catchment_area: gpd.GeoDataFrame) -> Non
...
+    else:
+        # Read roof surface polygons from outside
+        log.info(f"Reading roof surface polygons file {EnvVariable.ROOF_SURFACE_DATASET_PATH}.")
+        roof_surface_polygons = gpd.read_file(EnvVariable.ROOF_SURFACE_DATASET_PATH, layer="BuildingPolygons")
+        # Store the building_point_data to the database table
+        log.info("Adding roof_surface_polygons table to the database.")
+        roof_surface_polygons.to_postgis("roof_surface_polygons", engine, index=False, if_exists="replace")
+
+
+def save_roof_surface_type_data_to_db(engine: Engine) -> None:
```
Comment (caseyli92) [inline src/digitaltwin/data_to_db.py:484]: avoid confusion

```suggestion
def save_roof_surface_data_to_db(engine: Engine) -> None:
```

File: src/digitaltwin/retrieve_static_boundaries.py [hunk 13]
```diff
@@ -44,6 +44,8 @@ def main(
     data_to_db.store_geospatial_layers_data_to_db(engine, catchment_area)
     # Store user log information in the database
     data_to_db.user_log_info_to_db(engine, catchment_area)
+    # Read roof_surface file and store in the database
+    data_to_db.save_roof_surface_type_data_to_db(engine)
```
Comment (caseyli92) [inline src/digitaltwin/retrieve_static_boundaries.py:48]:
```suggestion
    data_to_db.save_roof_surface_data_to_db(engine)
```

File: src/pollution_model/run_medusa_2.py
Comment (LukeParky) [inline src/pollution_model/run_medusa_2.py:1]: Add descriptions for tests

**Review by caseyli92 — CHANGES_REQUESTED** [review]
Great work! I have a few suggestions and comments below. Also, please remember to add a short description to each of the Medusa unit tests.

File: Dockerfile [hunk 14]
```diff
@@ -49,6 +49,11 @@ RUN mkdir /stored_data \
     && setfacl -R -m u:nonroot:rwx /stored_data \
     && mkdir /stored_data/geoserver
 
+# Add clipped roof surfaces data into image
+COPY --chown=root:root --chmod=444 \
```
[the remark this answers is not in the document]
Comment (LukeParky) [reply Dockerfile:53]: The directory is created during copy by this command: https://github.com/GeospatialResearch/Digital-Twins/blob/aa6ba2363fa0e3eedb3e6acbf6ececd67c028e2b/Dockerfile#L52-L55 so the `roof_surfaces_data` directory does not need to be created.

The road surfaces aren't included in this PR so I will leave that for #286.

File: docker-compose-dev.yml [hunk 15]
```diff
@@ -20,6 +20,13 @@ services:
       - ${DATA_DIR}:/stored_data
       - ${DATA_DIR_MODEL_OUTPUT}:/stored_data/model_output
       - ${DATA_DIR_GEOSERVER}:/stored_data/geoserver
+      - ${ROOF_SURFACE_DATASET_PATH}:/stored_data/roof_surfaces_data/roof_surfaces.gdb
+
```
[the remark this answers is not in the document]
Comment (LukeParky) [reply docker-compose-dev.yml:24]: Deferring to #286, see comment https://github.com/GeospatialResearch/Digital-Twins/pull/281/files#r1845696839

**Review by caseyli92 — APPROVED** [review]

**Pull request merged**